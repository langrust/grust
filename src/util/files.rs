use codespan_reporting::files;
use std::ops::Range;

/// A file that is backed by an `Arc<String>`.
#[derive(Debug, Clone)]
struct File {
    /// The name of the file.
    name: String,
    /// The source code of the file.
    source: String,
    /// The starting byte indices in the source code.
    line_starts: Vec<usize>,
}

impl File {
    fn line_start(&self, line_index: usize) -> Result<usize, files::Error> {
        use std::cmp::Ordering;

        match line_index.cmp(&self.line_starts.len()) {
            Ordering::Less => Ok(*self
                .line_starts
                .get(line_index)
                .expect("failed despite previous check")),
            Ordering::Equal => Ok(self.source.len()),
            Ordering::Greater => Err(files::Error::LineTooLarge {
                given: line_index,
                max: self.line_starts.len() - 1,
            }),
        }
    }
}

/// An opaque file identifier.
#[derive(Copy, Clone, PartialEq, Eq, Debug)]
pub struct FileId(i32);
impl FileId {
    /// Construct and return the default file identifier.
    pub fn default() -> Self {
        return FileId(-1);
    }
}

/// The files handler.
#[derive(Debug, Clone)]
pub struct Files {
    files: Vec<File>,
}

impl Files {
    /// Create a new files database.
    pub fn new() -> Files {
        Files { files: Vec::new() }
    }

    /// Add a file to the database, returning the handle that can be used to
    /// refer to it again.
    pub fn add(
        &mut self,
        name: impl Into<String>,
        source: impl Into<String>,
    ) -> Option<FileId> {
        let file_id = FileId(i32::try_from(self.files.len()).ok()?);
        let name = name.into();
        let source = source.into();
        let line_starts = files::line_starts(&source).collect();

        self.files.push(File {
            name,
            line_starts,
            source,
        });

        Some(file_id)
    }

    /// Get the file corresponding to the given id.
    fn get(&self, file_id: FileId) -> Result<&File, files::Error> {
        self.files
            .get(file_id.0 as usize)
            .ok_or(files::Error::FileMissing)
    }
}

impl<'files> files::Files<'files> for Files {
    type FileId = FileId;
    type Name = &'files str;
    type Source = &'files str;

    fn name(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(self.get(file_id)?.name.as_ref())
    }

    fn source(&self, file_id: FileId) -> Result<&str, files::Error> {
        Ok(&self.get(file_id)?.source)
    }

    fn line_index(
        &self,
        file_id: FileId,
        byte_index: usize
    ) -> Result<usize, files::Error> {
        self.get(file_id)?
            .line_starts
            .binary_search(&byte_index)
            .or_else(|next_line| Ok(next_line - 1))
    }

    fn line_range(
        &self,
        file_id: FileId,
        line_index: usize,
    ) -> Result<Range<usize>, files::Error> {
        let file = self.get(file_id)?;
        let line_start = file.line_start(line_index)?;
        let next_line_start = file.line_start(line_index + 1)?;

        Ok(line_start..next_line_start)
    }
}