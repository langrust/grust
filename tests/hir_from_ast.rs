use codespan_reporting::files::SimpleFiles;

use grustine::{error::display, hir_from_ast};

#[test]
fn hir_from_ast_transformation_for_counter() {
    let mut files = SimpleFiles::new();

    let counter_id = files.add(
        "counter.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/counter.gr").expect("unkown file"),
    );

    match hir_from_ast(counter_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_blinking() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "blinking.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/blinking.gr").expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/button_management.gr").expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management_condition_match() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management_condition_match.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/button_management_condition_match.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_button_management_using_function() {
    let mut files = SimpleFiles::new();

    let blinking_id = files.add(
        "button_management_using_function.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/button_management_using_function.gr")
            .expect("unkown file"),
    );

    match hir_from_ast(blinking_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}

#[test]
fn hir_from_ast_transformation_for_pid() {
    let mut files = SimpleFiles::new();

    let pid_id = files.add(
        "pid.gr",
        std::fs::read_to_string("tests/fixture/hir_from_ast/pid.gr").expect("unkown file"),
    );

    match hir_from_ast(pid_id, &mut files) {
        Ok(file) => insta::assert_yaml_snapshot!(file),
        Err(errors) => display(&errors, &files),
    }
}
