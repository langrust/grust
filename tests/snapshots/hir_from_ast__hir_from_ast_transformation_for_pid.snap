---
source: tests/hir_from_ast.rs
expression: file
---
typedefs:
  - Structure:
      id: GainPID
      fields:
        - - k_p
          - Float
        - - k_i
          - Float
        - - k_d
          - Float
      location: ~
functions:
  - id: min
    inputs:
      - - a
        - Float
      - - b
        - Float
    statements:
      - id: test
        element_type: Boolean
        expression:
          Application:
            function_expression:
              Call:
                id: " > "
                typing:
                  Abstract:
                    - - Float
                      - Float
                    - Boolean
                location: ~
            inputs:
              - Call:
                  id: a
                  typing: Float
                  location: ~
              - Call:
                  id: b
                  typing: Float
                  location: ~
            typing: Boolean
            location: ~
        location: ~
      - id: result
        element_type: Float
        expression:
          Application:
            function_expression:
              Call:
                id: if_then_else
                typing:
                  Abstract:
                    - - Boolean
                      - Float
                      - Float
                    - Float
                location: ~
            inputs:
              - Call:
                  id: test
                  typing: Boolean
                  location: ~
              - Call:
                  id: b
                  typing: Float
                  location: ~
              - Call:
                  id: a
                  typing: Float
                  location: ~
            typing: Float
            location: ~
        location: ~
    returned:
      - Float
      - Call:
          id: result
          typing: Float
          location: ~
    location: ~
nodes:
  - id: integrate
    is_component: false
    inputs:
      - - x
        - Float
      - - dt
        - Float
    unscheduled_equations:
      MAX_INTEGRALE:
        scope: Local
        id: MAX_INTEGRALE
        signal_type: Float
        expression:
          Constant:
            constant:
              Float: 1000
            typing: Float
            location: ~
            dependencies: ~
        location: ~
      i:
        scope: Output
        id: i
        signal_type: Float
        expression:
          FunctionApplication:
            function_expression:
              Call:
                id: min
                typing:
                  Abstract:
                    - - Float
                      - Float
                    - Float
                location: ~
            inputs:
              - FunctionApplication:
                  function_expression:
                    Call:
                      id: " + "
                      typing:
                        Abstract:
                          - - Float
                            - Float
                          - Float
                      location: ~
                  inputs:
                    - SignalCall:
                        signal:
                          id: prev_i
                          scope: Local
                        typing: Float
                        location: ~
                        dependencies: ~
                    - FunctionApplication:
                        function_expression:
                          Call:
                            id: " * "
                            typing:
                              Abstract:
                                - - Float
                                  - Float
                                - Float
                            location: ~
                        inputs:
                          - SignalCall:
                              signal:
                                id: x
                                scope: Input
                              typing: Float
                              location: ~
                              dependencies: ~
                          - SignalCall:
                              signal:
                                id: dt
                                scope: Input
                              typing: Float
                              location: ~
                              dependencies: ~
                        typing: Float
                        location: ~
                        dependencies: ~
                  typing: Float
                  location: ~
                  dependencies: ~
              - SignalCall:
                  signal:
                    id: MAX_INTEGRALE
                    scope: Local
                  typing: Float
                  location: ~
                  dependencies: ~
            typing: Float
            location: ~
            dependencies: ~
        location: ~
      prev_i:
        scope: Local
        id: prev_i
        signal_type: Float
        expression:
          FollowedBy:
            constant:
              Float: 0
            expression:
              SignalCall:
                signal:
                  id: i
                  scope: Output
                typing: Float
                location: ~
                dependencies: ~
            typing: Float
            location: ~
            dependencies: ~
        location: ~
    unitary_nodes: {}
    location: ~
    graph: ~
  - id: derive
    is_component: false
    inputs:
      - - x
        - Float
      - - dt
        - Float
    unscheduled_equations:
      d:
        scope: Output
        id: d
        signal_type: Float
        expression:
          FunctionApplication:
            function_expression:
              Call:
                id: " / "
                typing:
                  Abstract:
                    - - Float
                      - Float
                    - Float
                location: ~
            inputs:
              - FunctionApplication:
                  function_expression:
                    Call:
                      id: (_)
                      typing:
                        Abstract:
                          - - Float
                          - Float
                      location: ~
                  inputs:
                    - FunctionApplication:
                        function_expression:
                          Call:
                            id: " - "
                            typing:
                              Abstract:
                                - - Float
                                  - Float
                                - Float
                            location: ~
                        inputs:
                          - SignalCall:
                              signal:
                                id: x
                                scope: Input
                              typing: Float
                              location: ~
                              dependencies: ~
                          - SignalCall:
                              signal:
                                id: prev_x
                                scope: Local
                              typing: Float
                              location: ~
                              dependencies: ~
                        typing: Float
                        location: ~
                        dependencies: ~
                  typing: Float
                  location: ~
                  dependencies: ~
              - SignalCall:
                  signal:
                    id: dt
                    scope: Input
                  typing: Float
                  location: ~
                  dependencies: ~
            typing: Float
            location: ~
            dependencies: ~
        location: ~
      prev_x:
        scope: Local
        id: prev_x
        signal_type: Float
        expression:
          FollowedBy:
            constant:
              Float: 0
            expression:
              SignalCall:
                signal:
                  id: x
                  scope: Input
                typing: Float
                location: ~
                dependencies: ~
            typing: Float
            location: ~
            dependencies: ~
        location: ~
    unitary_nodes: {}
    location: ~
    graph: ~
component:
  id: pid
  is_component: true
  inputs:
    - - v_c
      - Float
    - - v
      - Float
    - - dt
      - Float
  unscheduled_equations:
    e:
      scope: Local
      id: e
      signal_type: Float
      expression:
        FunctionApplication:
          function_expression:
            Call:
              id: " - "
              typing:
                Abstract:
                  - - Float
                    - Float
                  - Float
              location: ~
          inputs:
            - SignalCall:
                signal:
                  id: v_c
                  scope: Input
                typing: Float
                location: ~
                dependencies: ~
            - SignalCall:
                signal:
                  id: v
                  scope: Input
                typing: Float
                location: ~
                dependencies: ~
          typing: Float
          location: ~
          dependencies: ~
      location: ~
    e_d:
      scope: Local
      id: e_d
      signal_type: Float
      expression:
        NodeApplication:
          node: derive
          inputs:
            - SignalCall:
                signal:
                  id: e
                  scope: Local
                typing: Float
                location: ~
                dependencies: ~
            - SignalCall:
                signal:
                  id: dt
                  scope: Input
                typing: Float
                location: ~
                dependencies: ~
          signal: d
          typing: Float
          location: ~
          dependencies: ~
      location: ~
    e_i:
      scope: Local
      id: e_i
      signal_type: Float
      expression:
        NodeApplication:
          node: integrate
          inputs:
            - SignalCall:
                signal:
                  id: e
                  scope: Local
                typing: Float
                location: ~
                dependencies: ~
            - SignalCall:
                signal:
                  id: dt
                  scope: Input
                typing: Float
                location: ~
                dependencies: ~
          signal: i
          typing: Float
          location: ~
          dependencies: ~
      location: ~
    gain:
      scope: Local
      id: gain
      signal_type:
        Structure: GainPID
      expression:
        Structure:
          name: GainPID
          fields:
            - - k_p
              - Constant:
                  constant:
                    Float: 0.2
                  typing: Float
                  location: ~
                  dependencies: ~
            - - k_i
              - Constant:
                  constant:
                    Float: 1.5
                  typing: Float
                  location: ~
                  dependencies: ~
            - - k_d
              - Constant:
                  constant:
                    Float: 6
                  typing: Float
                  location: ~
                  dependencies: ~
          typing:
            Structure: GainPID
          location: ~
          dependencies: ~
      location: ~
    u:
      scope: Output
      id: u
      signal_type: Float
      expression:
        Match:
          expression:
            SignalCall:
              signal:
                id: gain
                scope: Local
              typing:
                Structure: GainPID
              location: ~
              dependencies: ~
          arms:
            - - Structure:
                  name: GainPID
                  fields:
                    - - k_p
                      - Identifier:
                          name: k_p
                          location: ~
                    - - k_i
                      - Identifier:
                          name: k_i
                          location: ~
                    - - k_d
                      - Identifier:
                          name: k_d
                          location: ~
                  location: ~
              - ~
              - []
              - FunctionApplication:
                  function_expression:
                    Call:
                      id: " + "
                      typing:
                        Abstract:
                          - - Float
                            - Float
                          - Float
                      location: ~
                  inputs:
                    - FunctionApplication:
                        function_expression:
                          Call:
                            id: " + "
                            typing:
                              Abstract:
                                - - Float
                                  - Float
                                - Float
                            location: ~
                        inputs:
                          - FunctionApplication:
                              function_expression:
                                Call:
                                  id: " * "
                                  typing:
                                    Abstract:
                                      - - Float
                                        - Float
                                      - Float
                                  location: ~
                              inputs:
                                - SignalCall:
                                    signal:
                                      id: k_p
                                      scope: Local
                                    typing: Float
                                    location: ~
                                    dependencies: ~
                                - SignalCall:
                                    signal:
                                      id: e
                                      scope: Local
                                    typing: Float
                                    location: ~
                                    dependencies: ~
                              typing: Float
                              location: ~
                              dependencies: ~
                          - FunctionApplication:
                              function_expression:
                                Call:
                                  id: " * "
                                  typing:
                                    Abstract:
                                      - - Float
                                        - Float
                                      - Float
                                  location: ~
                              inputs:
                                - SignalCall:
                                    signal:
                                      id: k_i
                                      scope: Local
                                    typing: Float
                                    location: ~
                                    dependencies: ~
                                - SignalCall:
                                    signal:
                                      id: e_i
                                      scope: Local
                                    typing: Float
                                    location: ~
                                    dependencies: ~
                              typing: Float
                              location: ~
                              dependencies: ~
                        typing: Float
                        location: ~
                        dependencies: ~
                    - FunctionApplication:
                        function_expression:
                          Call:
                            id: " * "
                            typing:
                              Abstract:
                                - - Float
                                  - Float
                                - Float
                            location: ~
                        inputs:
                          - SignalCall:
                              signal:
                                id: k_d
                                scope: Local
                              typing: Float
                              location: ~
                              dependencies: ~
                          - SignalCall:
                              signal:
                                id: e_d
                                scope: Local
                              typing: Float
                              location: ~
                              dependencies: ~
                        typing: Float
                        location: ~
                        dependencies: ~
                  typing: Float
                  location: ~
                  dependencies: ~
          typing: Float
          location: ~
          dependencies: ~
      location: ~
  unitary_nodes: {}
  location: ~
  graph: ~
location: ~

